drivers/iommu/riscv/io_pgtable.c

```diff
@@ -104,6 +105,7 @@ static int riscv_iommu_map_pages(struct io_pgtable_ops *ops,
 		pte_val = pfn_pte(phys_to_pfn(phys),
 				  (prot & IOMMU_WRITE) ? PAGE_WRITE :
 				  PAGE_READ);
+		pte_val = pte_mkdirty(pte_val);
```

Comment:
I've pushed internal fix for A/D bit updates based on requested mapping permission modes (as allowed by the priv spec).  Please have a look at https://github.com/tjeznach/linux/commit/c8a002e28f49930f0a9b8c626ab34d7cb24bf985 and let me know if that works.

Reply:
ACK, I'll re-base the patch on top of  your own changes.